### ISSUE TITLE
Block elements get replaced when wrapping them in a custom multiline container blot with block scope

### ISSUE THREAD

**Issue by mbv-og** [quilljs/parchment#34]
I wrote a custom blot that extends blot/container to wrap several lines into a collapsible element (something that can e.g. be toggled via the bootstrap collapse functionality). I can create such a container and place any other blots inside, embeds, inlines and blocks alike. Copy & paste also causes no problems.

The issue I have occurs when I try to select multiple lines and turn those into a collapsible section. Inline-blots like bold and italic remain unchanged, however any block blot gets completely replaced by the blot type I have set as _defaultChild_ in my class. I assume that the problem here is that I have to set the scope as _Parchment.Scope.BLOCK_BLOT_ for it to work. I have traced the issue to [this line in the FormatBlot class](https://github.com/quilljs/parchment/blob/master/src/blot/abstract/format.ts#L32), where it replaces one block element with another (probably to prevent nesting of certain block elements that shouldn't allow it?)

So I was wondering if it might be a good idea to add another If there, checking whether the new blot is an instance of Container, and if so use **this.wrap** instead of **this.replaceWith**. Would it be easily possible to add a check like that, or might that break other parts of the editor?

**Comment by jhchen** [comment]
All BlockBlots are ContainerBlots so the check would always pass if that were changed. This is a bit of an advanced topic so docs are still forthcoming but I'd take a look at how Quill implements Lists as an example of how you might achieve what you are going for.

**Comment by mbv-og** [comment]
Yeah, I downloaded the non-minified version of Quill and did some more testing and realized it's not that easy after all.  In the end I was able to trace one half of the problem to the replace-Method of my Collapsible-Blot.  Instead of the original line I got from the List-Blot..
```es6
let item = Parchment.create(this.statics.defaultChild);
```
..I had to use this to retain the inner formatting of the children:
```es6
let item;

if (this.statics.allowedChildren.indexOf(target.constructor) < 0) {
	item = target.clone();
}
else {
	item = Parchment.create(this.statics.defaultChild);
}
```

That worked for most formats, except for the List (and probably Code, but I didn't test that because that is not a format we allow anyway).  To fix that, I ultimately altered an If in the replaceWith Method of the ListItem (in the compiled code for now, but I will fork the List and replace the format with my own version).

https://github.com/quilljs/quill/blob/develop/formats/list.js#L29  had to be changed to this to work for me (from my Test in the compiled version of the code):
```javascript
var newParent = Registry.query(name);
if (name === this.parent.statics.blotName || newParent.__proto__.name === _container.default.name) {
```
That allowed me to do something like this:
![image](https://user-images.githubusercontent.com/4964032/28007427-a930a0e8-6553-11e7-91a2-8266ae982251.png)

But making the example for the screenshot above I just found a bug I introduced by doing that; I cannot remove a bullet point of an empty list item line using Backspace anymore. Back to debugging.

**Comment by mbv-og** [comment]
I got it to work with just a little tweak to my copy of the ListItem class; changing the line of the replaceWith method (https://github.com/quilljs/quill/blob/develop/formats/list.js#L29 in the original file)
```es6
if (name === this.parent.statics.blotName) {
```
in the following way:
```es6
var QuillBlock = Quill.import('blots/block');
var Parchment = Quill.import('parchment');


class ListItem extends QuillBlock
{
    static formats(domNode) {
        return domNode.tagName === this.tagName ? undefined : super.formats(domNode);
    }

    format(name, value) {
        if (name === Quill.import('formats/list').blotName && !value) {
            this.replaceWith(Parchment.create(this.statics.scope));
        }
        else {
            super.format(name, value);
        }
    }

    remove() {
        if (this.prev == null && this.next == null) {
            this.parent.remove();
        }
        else {
            super.remove();
        }
    }

    replaceWith(name, value) {
        const newParent = Parchment.query(name);
        const currentParent = this.parent;

        currentParent.isolate(this.offset(currentParent), this.length());

        if (
            name === this.parent.statics.blotName
            || (
                null !== newParent
                && newParent.allowedChildren.some(function (child) {
                    return (
                        child.prototype.statics.scope === Parchment.Scope.BLOCK_BLOT
                        && currentParent instanceof child
                    );
                })
            )
        ) {
            this.parent.replaceWith(name, value);
            return this;
        }
        else {
            this.parent.unwrap();
            return super.replaceWith(name, value);
        }
    }
}

ListItem.blotName = 'list-item';
ListItem.tagName = 'LI';

Quill.register('formats/list/item', ListItem, true);
```
allows my collapsible text container to retain the List as long as I have it added to the allowedChildren (which I had done already as my very first attempt to make it work anyway). It also only works if the new parent is a Block element so it's still valid html.

I guess with that the issue can be closed, and it turned out to be a Quill formats thing, not Parchment. Sorry about that.

**Comment by jhchen** [comment]
Not sure I followed everything but thanks for sharing.

**Comment by johnking** [comment]
Hi, @mbv-og 

  May you please share me how to use your customized ListItem in your high level code? , i.e., how to use/insert some ListItem into the quill editor? 

  There are only `insertText` and `insertEmbed` APIs available from Quill. All the quill examples available are about customizing embed blot.  

  Here is my question posted at Quill forum: [How to customize a sentenceBlot/paragraphBlot and insert it into Quill?](https://github.com/quilljs/quill/issues/1581)

  Basically, I don't know how to use the customized blot based on block in quill editor. 

thanks for your help.

-John

**Comment by davidmold** [comment]
If you want a blot based on Blot to be able to contain arbitrary other blots, you have to set its allowedChildren field to null, e.g.

`
class MyBlot extends Block {
}
MyBlot.tagName = 'DIV'
MyBlot.className = 'special-div'
MyBlot.allowedChildren = null
`
